Week 4 PDF

rojenarda opened the pull request.

novruzamirov reviewed (changes requested):
1) At slide 4, "Constant time (O(1)) insertion/deletion.". It is not true for Singly Linked List, in Singly Linked List the insertion/deletion complexity is O(N). Nevertheless, in doubly linked list, it is O(1). 
2) At slide 26, there is a problem on "Output:" part. Fix the "???".
3) At slide 27, "struct Comparator{" part can not be seen, if you added it to the next slide, delete the one in slide 27, because it is seen half.
4) At slide 32, "ancestor" if "a = b" is wrong i think. The ancestor means, if it is the parent of the node or more higher level. Therefore, I do not think the part of sentence "Node a is ancestor of node b if a = b or" is correct
5) At slide 32, "descendant" is not correct either, it must be "Node b is descendant of node a if a is an ancestor of b"

rojenarda commented:
1. My emphasis was that if we have the pointer to the node, deleting it or inserting a new node takes O(1) time. Acquiring the pointer (random access) takes O(n) time. I didn't change the file but I will explain this clearly in the class to avoid any confusions.

rojenarda commented:
1. My emphasis was that if we have the pointer to the node, deleting it or inserting a new node takes O(1) time. Acquiring the pointer (random access) takes O(n) time. I didn't change the file but I will explain this clearly in the class to avoid any confusions.
2. Fixed slide 26.
3. Fixed slide 27.
4. "A node u is an ancestor of a node v if u = v or u is an ancestor of the parent of v. Conversely, we say that a node v is a descendent of a node u if u is an ancestor of v." (Michael T. Goodrich, Roberto Tamassia, David M. Mount - Data Structures and Algorithms in C++ -Wiley (2011), page 270)
5. Fixed slide 32. 
    Thank you.
